### PR TITLE
Implement transaction consumption

### PR DESCRIPTION
**Pull request by ripzery**
Issue/Task Number: 219-implement-transactionconsumption

# Overview

**TL;DR**

omg

**TU;DU** (Too Ugly Don't Understand)

This PR implemented the transaction consumption which will be used to consume the generated transaction ([this pull request](https://github.com/omisego/android-sdk/pull/19)). The consumer can *send* or *receive* the token *to* or *from* the transaction creator. To make the transaction flow **complete**, it takes mainly **4** steps as the following.

1. ~Generate a transaction (done)~
2. **Consume a transaction (this pr)**
3. *Approve or Reject a transaction (coming soon)*
4. *Implement websocket to receive a notification when the transaction is already completed (coming next month)*

# Changes

- Added consume the transaction request function to the SDK

# Implementation Details

1. Added `TransactionConsumptionParams` model to **send** to the `/me.consume_transaction_request` API

**Note**: If `the amount` and `the amount from the transaction request` was not specified, then `IllegalArgumentException` will be thrown with the message **"The transactionRequest amount or the amount of minted token to transfer should be provided"**

2. Added more params to `TransactionRequestCreateParams`

**Note**: If `allowAmountOverride` is `false` and `amount` is null, then `IllegalArgumentException` will be thrown with the message **""The amount cannot be null if the allowAmountOverride is false""**

3. Added `TransactionConsumption` model to **receive** the `/me.consume_transaction_request` API

4. Added an extension function `TransactionRequest.toTransactionConsumptionParams` to convert the `TransactionRequest` to `TransactionConsumptionParams` conveniently.

5. Added function `consumeTransactionRequest` to send the `TransactionConsumptionParams` to the server.

6. Add section `Consume a transaction request` to `README.md` and refactor the content to be more reasonable.

7. Added test to cover the initialization step of `TransactionConsumptionParams` and `TransactionRequestCreateParams`to ensure it can't be misuse.

# Usage

```kotlin
/* Short version */
val request = TransactionConsumptionParams.create(
    transactionRequest
)

/* Full version */
val request = TransactionConsumptionParams.create(
    transactionRequest,
    amount = 25_000.bd, // BigDecimal
    address = "an address",
    tokenId = "A token id",
    idempotencyToken = "An idempotency token",
    correlationId = "a correlation id",
    metadata = mapOf(),
    encryptedMetadata = mapOf()
)

omgAPIClient.consumeTransactionRequest(request).enqueue(object : OMGCallback<TransactionConsumption> {
    override fun success(response: OMGResponse<TransactionConsumption>) {
        // Handle success
    }

    override fun fail(response: OMGResponse<APIError>) {
        // Handle error
    }
})
```

# Impact

`N/A`

### CODE REVIEW

File: omisego-sdk/src/test/kotlin/co/omisego/omisego/model/transaction/request/TransactionRequestCreateParamsTest.kt [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+ */
+
+import co.omisego.omisego.extension.bd
+import org.amshove.kluent.shouldThrow
+import org.amshove.kluent.withMessage
+import org.junit.Test
+
+class TransactionRequestCreateParamsTest {
+    @Test
+    fun `TransactionRequestCreateParams should throws IllegalArgumentException if allowAmountOverride is false and the amount is null`() {
```
Comment (mederic-p) [inline omisego-sdk/src/test/kotlin/co/omisego/omisego/model/transaction/request/TransactionRequestCreateParamsTest.kt]: throws -> throw

File: omisego-sdk/src/test/kotlin/co/omisego/omisego/model/transaction/request/TransactionRequestCreateParamsTest.kt [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    fun `TransactionRequestCreateParams should be created successfully if the allowAmountOverride is false and the amount is not null`() {
+        TransactionRequestCreateParams(
+            TransactionRequestType.RECEIVE,
+            "OMG-1234-5678",
+            100.bd
+        )
+    }
+
+    @Test
+    fun `TransactionRequestCreateParams should be able to created if allowAmountOverride is true and the amount is null`() {
```
Comment (mederic-p) [inline omisego-sdk/src/test/kotlin/co/omisego/omisego/model/transaction/request/TransactionRequestCreateParamsTest.kt]: to created -> to be created

File: omisego-sdk/src/test/kotlin/co/omisego/omisego/model/transaction/request/TransactionRequestCreateParamsTest.kt [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        TransactionRequestCreateParams(
+            TransactionRequestType.RECEIVE,
+            "1234",
+            null,
+            allowAmountOverride = true
+        )
+    }
+
+    @Test
+    fun `TransactionRequestCreateParams should be able to created if allowAmountOverride is true and the amount is not null`() {
```
Comment (mederic-p) [inline omisego-sdk/src/test/kotlin/co/omisego/omisego/model/transaction/request/TransactionRequestCreateParamsTest.kt]: to created -> to be created

File: omisego-sdk/src/main/java/co/omisego/omisego/extension/BigDecimal.kt [hunk 4]
```diff
@@ -14,4 +14,7 @@ import java.math.BigDecimal
  *  For example, 100.0.bd is equal to BigDecimal.valueOf(100.0)
  */
 val Double.bd: BigDecimal
-    get() = BigDecimal.valueOf(this)
\ No newline at end of file
+    get() = BigDecimal.valueOf(this)
+
+val Int.bd: BigDecimal
+    get() = this.toBigDecimal()
```
Comment (badoualy) [inline omisego-sdk/src/main/java/co/omisego/omisego/extension/BigDecimal.kt]: `this` is probably not needed, not a big deal though :)

File: omisego-sdk/src/main/java/co/omisego/omisego/extension/Density.kt [hunk 5]
```diff
@@ -16,4 +15,4 @@ import android.content.res.Resources
  * This simplifies code dealing with dimensions.
  */
 inline val Int.dp: Float
-    get() = (this * Resources.getSystem().displayMetrics.density)
\ No newline at end of file
+    get() = (this * Resources.getSystem().displayMetrics.density)
```
Comment (badoualy) [inline omisego-sdk/src/main/java/co/omisego/omisego/extension/Density.kt]: Need to be careful with this, value could be wrong when using multiple screens :)

File: omisego-sdk/src/main/java/co/omisego/omisego/model/transaction/request/TransactionRequest.kt [hunk 6]
```diff
@@ -7,17 +7,157 @@ package co.omisego.omisego.model.transaction.request
...
+ * In the case of a type "send", this will be the token that the consumer will receive
+ * In the case of a type "receive" this will be the token that the consumer will send
+ * @param idempotencyToken The idempotency token to use for the consumption
+ * @param correlationId An id that can uniquely identify a transaction. Typically an order id from a provider.
+ * @param metadata Additional metadata for the consumption
+ * @param encryptedMetadata Additional encrypted metadata for the consumption
+ *
+ * @return The [TransactionConsumptionParams] used for consume the a transaction request
+ */
+fun TransactionRequest.toTransactionConsumptionParams(
```
Comment (badoualy) [inline omisego-sdk/src/main/java/co/omisego/omisego/model/transaction/request/TransactionRequest.kt]: If this file starts to get too big, moving this in a `converter` package could be a good idea. Not an issue for now 😄

File: omisego-sdk/src/main/java/co/omisego/omisego/model/transaction/request/TransactionRequestCreateParams.kt [hunk 7]
```diff
@@ -1,32 +1,35 @@
...
+
 /**
  * Represents a structure used to generate a transaction request
  *
  */
 data class TransactionRequestCreateParams(
 
     /**
-     * The type of transaction to be generated
+     * The type of transaction to be generated (send of receive)
```
Comment (badoualy) [inline omisego-sdk/src/main/java/co/omisego/omisego/model/transaction/request/TransactionRequestCreateParams.kt]: `or` ?

File: omisego-sdk/src/main/java/co/omisego/omisego/network/ewallet/EWalletAPI.kt [hunk 8]
```diff
@@ -49,4 +53,10 @@ interface EWalletAPI {
 
     @POST(RETRIEVE_TRANSACTION_REQUEST)
     fun retrieveTransactionRequest(@Body request: TransactionRequestParams): OMGCall<TransactionRequest>
+
+    @POST(Endpoints.CONSUME_TRANSACTION_REQUEST)
```
Comment (badoualy) [inline omisego-sdk/src/main/java/co/omisego/omisego/network/ewallet/EWalletAPI.kt]: You used static import for other constants :)

File: omisego-sdk/src/main/java/co/omisego/omisego/model/transaction/request/TransactionRequest.kt (hunk 6 above)
Comment (ripzery) [reply omisego-sdk/src/main/java/co/omisego/omisego/model/transaction/request/TransactionRequest.kt]: Good idea 👍

File: omisego-sdk/src/main/java/co/omisego/omisego/model/transaction/request/TransactionRequestCreateParams.kt (hunk 7 above)
Comment (ripzery) [reply omisego-sdk/src/main/java/co/omisego/omisego/model/transaction/request/TransactionRequestCreateParams.kt]: Oh, you're right 👍

File: omisego-sdk/src/main/java/co/omisego/omisego/network/ewallet/EWalletAPI.kt (hunk 8 above)
Comment (ripzery) [reply omisego-sdk/src/main/java/co/omisego/omisego/network/ewallet/EWalletAPI.kt]: I've already fixed it in the future PR (can't remember which one) 👍
